File: apps/ayokoding-www/content/en/learn/courses/type-systems/learning/capstone/code/Domain.hs
module Domain where

newtype Email = Email String
data Verification = Unverified Email | Verified Email

parse :: String -> Either String Email
parse raw | '@' `elem` raw = Right (Email raw)
          | otherwise = Left "email must contain @"

verify :: Email -> Verification
verify = Verified

send :: Verification -> String
send (Verified (Email value)) = "sent to " ++ value
send (Unverified _) = "not verified"

File: apps/ayokoding-www/content/en/learn/courses/type-systems/learning/capstone/code/domain.ml
type email = Email of string
type verification = Unverified of email | Verified of email

let parse raw =
  if String.contains raw '@' then Ok (Email raw) else Error "email must contain @"

let verify = function
  | Email value -> Verified (Email value)

let send = function
  | Verified (Email value) -> "sent to " ^ value

let () =
  match parse "reader@example.test" with
  | Ok address -> print_endline (send (verify address))
  | Error message -> print_endline message
